code/modules/network/network.dm

```diff
@@ -0,0 +1,119 @@
+	var/stealth = 0									// Whether the network should be hidden from standard probing and prying.
+	var/wireless = 0								// Whether the network can be accessed remotely.
+	var/datum/network/root 							// The root network, if this is a subnetwork.
+	var/list/datum/network/sub = list() 			// A list of networks that are a subnetwork of this one.
+	var/list/datum/network_command/commands = list()	// A list of commands this network has access to. Works kinda like virus symptoms.
+
+/datum/network/New(atom/holder)
+	..()
+	src.holder = holder
+	if(id)
+		while(networks_by_id[id])
+			id = initial(id)+"_"+random_string(6, hex_characters)
+		networks_by_id[id] = src
+		active_network_ids += id
+		if(wireless)
+			networks_by_wide += src
+	var/list/commandlist = commands
+	commands = list()
+	for(var/C in commandlist)
+		commands += new C()
+
+/datum/network/Destroy()
+	holder = null
+	for(var/datum/network/S in sub)
+		S.del_root()
+	sub = list()
+	if(root)
+		root.del_sub(src)
+	root = list()
+	for(var/datum/network_command/NC in commands)
+		qdel(NC)
+	commands = list()
+	if(id)
+		networks_by_id[id] = null
+		networks_by_id -= id
+		active_network_ids -= id
+	if(wireless)
+		networks_by_wide -= src
+	return ..()
+
+/datum/network/proc/del_root() // Called when the root is destroyed.
+	root = null
+
+/datum/network/proc/del_sub(datum/network/S) // Called when the subdirectory is destroyed.
+	if(S in sub)
+		sub -= S
+
+/datum/network/proc/execute(var/command, var/obj/item/device/hacktool/H) // Used to pass commands to the network.
+	if(!command)
+		return
+	var/datum/network_argument/A = new(command)
+
+	var/list/data = list()
+	for(var/datum/network_command/NC in commands) // allows for adding 'universal' commands, which can be attached to any network object.
+		if(NC.trigger == A.main_command)
+			NC.execute(src, A, H) // Commands can be hard-coded to the network object, or made into a 'universal' command, which can be called by any network object.
+			qdel(A)
+			return 1
+	return
+
+/datum/network_command/proc/feed(var/obj/item/device/hacktool/H)
+	if(feedback && H)
+		H.feed(feedback)
+		return 1
+
+/datum/network_command/proc/disconnect(var/obj/item/device/hacktool/H)
+	if(H)
+		H.disconnect()
+
+/datum/network_command/proc/connect(var/datum/network/N, var/obj/item/device/hacktool/H)
+	if(H && N)
+		H.connect(N)
+
+
+/*
+	Intended to increase code-modularity and reuseability.
+	Execute should always return either null, or a list.
+*/
+
+/datum/network_command
+	var/trigger = "" 	// The trigger word for this command to be executed.
+	var/stealth = 0		// Whether this command can be identified using info.
+	var/feedback = ""	// The message sent back to the device calling this command.
+	var/info = ""		// A feedback message that explains the usage of this command. For the noobs.
+
+/datum/network_command/proc/execute(datum/network/N, datum/network_argument/A, obj/item/device/hacktool/H)
+	if(!N || !A)
```

Comment:
this proc does nothing anyway, so there's no point making sure it has valid data... to do nothing with.